frame/composable-traits/src/staking/mod.rs

```diff
@@ -1,35 +1,138 @@
+pub struct RewardPool<
+	AccountId,
+	AssetId,
+	Balance,
+	BlockNumber,
+	DurationPresets,
+	RewardsLength: Get<u32>,
+> {
+	pub owner: AccountId,
+
+	/// The staked asset id of the reward pool.
+	pub asset_id: AssetId,
+
+	/// rewards accumulated
+	pub rewards: BoundedBTreeMap<AssetId, Reward<AssetId, Balance>, RewardsLength>,
+
+	/// Total shares distributed among stakers
+	pub total_shares: Balance,
+
+	/// Pool would stop adding rewards to pool at this block number.
+	pub end_block: BlockNumber,
+
+	// possible lock config for this pool
+	pub lock: LockConfig<DurationPresets>,
+}
+
+/// Reward configurations for a given asset type.
+#[derive(RuntimeDebug, PartialEq, Eq, Clone, MaxEncodedLen, Encode, Decode, TypeInfo)]
+pub struct RewardConfig<AssetId, Balance> {
+	/// asset id of the reward
+	pub asset_id: AssetId,
+
+	/// Upper bound on the `total_rewards - total_dilution_adjustment`.
+	pub max_rewards: Balance,
+
+	/// The rewarding rate that increases the pool `total_reward` (and `actual_total_reward`)
+	/// at a given time.
+	pub reward_rate: Perbill,
+}
+
+/// Categorize the reward pool by it's incentive characteristics and expose
+/// initial configuration parameters.
+/// TODO refer to the relevant section in the design doc.
+#[derive(RuntimeDebug, Encode, Decode, MaxEncodedLen, Clone, PartialEq, Eq, TypeInfo)]
+pub enum RewardPoolConfiguration<AccountId, AssetId, Balance, BlockNumber, DurationPresets> {
+	/// A pool with an adjustable reward rate to be used as incentive.
+	RewardRateBasedIncentive {
+		/// Protocol or the user account that owns this pool
+		owner: AccountId,
+		/// The staked asset id of the reward pool.
+		asset_id: AssetId,
+		/// Pool would stop adding rewards to pool at this block number.
+		end_block: BlockNumber,
+		/// initial reward configuration for this pool
+		initial_reward_config: RewardConfig<AssetId, Balance>,
+		// possible lock config for this reward
+		lock: LockConfig<DurationPresets>,
+	},
+}
+
+/// Staking typed fNFT, usually can be mapped to raw fNFT storage type. A position identifier
+/// should exist for each position when stored in the runtime storage.
+/// TODO refer to the relevant section in the design doc.
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
+pub struct Stake<RewardPoolId, AssetId, Balance, RewardsLength: Get<u32>> {
+	/// Reward Pool ID from which pool to allocate rewards for this
+	pub reward_pool_id: RewardPoolId,
+
 	/// The original stake this NFT was minted for or updated NFT with increased stake amount.
-	pub real_stake: Balance,
-	/// List of reward asset/pending rewards.
-	pub rewards: Rewards,
-	pub lock: Option<Lock>,
-	/// The reward multiplier.
-	pub reward_multiplier: Perbill,
+	pub stake: Balance,
+
+	/// Pool share received for this NFT
+	pub share: Balance,
+
+	/// Reduced rewards by asset for the position (d_n)
+	pub reductions: BoundedBTreeMap<AssetId, Balance, RewardsLength>,
```

Comment:
this is not traits, not public API of pallet

Reply:
```suggestion
	reductions: BoundedBTreeMap<AssetId, Balance, RewardsLength>,
```